### PR TITLE
feat: adding resource support

### PR DESCRIPTION
**Pull request by aaron-steinfeld**
## Description
This change adds support to the data model to support propagating resources

### Testing
Avro compat check verifies backwards compatibility

Ref: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/sdk.md

### CODE REVIEW

File: data-model/src/main/avro/RawSpan.avdl [hunk 1]
```diff
@@ -12,6 +13,8 @@ protocol RawSpanProtocol {
 
     array<Entity> entity_list = [];
 
+    union { null, Resource } resource = null;
```
Comment (aaron-steinfeld) [inline data-model/src/main/avro/RawSpan.avdl:16]: Because span normalization right now works one span at a time (Ref: https://github.com/hypertrace/hypertrace-ingester/blob/296351d7b6b7b89061cf581f69c7fa0da83f46bf/span-normalizer/span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/jaeger/JaegerSpanToAvroRawSpanTransformer.java#L43 ) we're stuck flattening resources across each RawSpan at the moment. The StructuredTrace data model re-normalizes it so this limitation does not cause all spans later in the pipeline to bloat.

File: data-model/src/main/avro/Event.avdl [hunk 2]
```diff
@@ -34,6 +34,8 @@ protocol EventProtocol {
     // Event may be associated with multiple entities
     array<string> entity_id_list = [];
 
+    int resource_index = -1;
```
Comment (tim-mwangi) [inline data-model/src/main/avro/Event.avdl:37]: This is not an list of indices?

Comment (aaron-steinfeld) [reply data-model/src/main/avro/Event.avdl:37]: I initially modeled it that way, but a span should be associated with at most one resource.

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java [hunk 3]
```diff
@@ -420,10 +421,21 @@ public static StructuredTrace buildStructuredTraceFromRawSpans(List<RawSpan> raw
                                                                  String customerId,
                                                                  Timestamps timestamps) {
     Map<String, Entity> entityMap = new HashMap<>();
+    // Relying on insertion ordered keyset
+    LinkedHashMap<Resource, Integer> resourceIndexMap = new LinkedHashMap<>();
```
Comment (rish691) [inline data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java:425]: type can be Map interface

File: data-model/src/main/avro/Event.avdl (hunk 2 above)
Comment (rish691) [inline data-model/src/main/avro/Event.avdl:37]: why is this index needed?

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java (hunk 3 above)
Comment (aaron-steinfeld) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java:425]: This was intentional, as we are relying on the properties of a linkedhashmap, not a standard map - they're not interchangeable here (indicated by the comment above).

File: data-model/src/main/avro/Event.avdl (hunk 2 above)
Comment (aaron-steinfeld) [reply data-model/src/main/avro/Event.avdl:37]: This is the main point of the change. A span is associated with at most one resource - this index references that resource which is being held in the structured trace. Resources are the tracing equivalent to entities (which never should have been on the structured trace), but they're immutable.

Comment (rish691) [reply data-model/src/main/avro/Event.avdl:37]: Resource will be reported alongside span message in the platform?

Comment (aaron-steinfeld) [reply data-model/src/main/avro/Event.avdl:37]: Yes - see links to OLTP reporting format above (specifically, the ones in https://github.com/hypertrace/data-model/pull/16#discussion_r592805664 )

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java [hunk 4]
```diff
@@ -191,6 +191,7 @@ private StructuredTrace build() {
     builder.setEventEdgeList(new ArrayList<>());
     builder.setEntityEdgeList(new ArrayList<>());
     builder.setEntityEventEdgeList(new ArrayList<>());
+    builder.setResourceList(this.resourceList);
```
Comment (kotharironak) [inline data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java:194]: Being public method `buildStructuredTrace`, in this case, we may have resource_list, but resource_index can be default in Event. If, we are not using this method outside, except `buildStructuredTraceFromRawSpans`, can we consider making it `private`?

File: data-model/src/main/avro/Event.avdl (hunk 2 above)
Comment (kotharironak) [inline data-model/src/main/avro/Event.avdl:37]: Can we add a comment here?

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java (hunk 4 above)
Comment (aaron-steinfeld) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java:194]: Can you rephrase? I'm not really sure what you're asking. This method, `build` is private. `resourceList` part of StructuredTrace and is generated, along with the builder we're calling here.

File: data-model/src/main/avro/Event.avdl (hunk 2 above)
Comment (aaron-steinfeld) [reply data-model/src/main/avro/Event.avdl:38]: Done

File: data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java (hunk 4 above)
Comment (kotharironak) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java:194]: I was referring to `buildStructuredTrace` being public. So, in the below-calling order, we will have resurce_index as `-1` for the Event. 
```
StructuredTraceBuilder builder = new StructuredTraceBuilder(..., resourceList);
builder.buildStructuredTrace()
```
But, it seems to me that `buildStructuredTrace` is not used outside of this class. So, was suggesting, if we can make that `private`.

Comment (aaron-steinfeld) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java:194]: I think this gets at how it's meant to be used. The existing design of this feels problematic for a number of reasons, but I was trying to work within it. The structured trace builder takes in events but does not mutate them - given that (which seems right to me), your scenario would work correctly. The problem in your scenario would be if we used the constructor with events that haven't been associated with resources yet. We'd have the same problem in other misconfigurations.

As that's the only non-private method on the builder, making it private is equivalent to making the whole class private and just exposing a public static util method. Is that what you're suggesting? I don't even think we should have this class at all to be honest, so I'm fine with that to start limiting it's usages.

Comment (kotharironak) [reply data-model/src/main/java/org/hypertrace/core/datamodel/shared/trace/StructuredTraceBuilder.java:194]: I agree with you about the need for this class. Yeah, I was thinking of exposing only the util method -`buildStructuredTraceFromRawSpans` which is only getting used in the HT side and restricting its usage. But, overall I am fine with model design.

**Review by kotharironak — APPROVED** [review]

**Review by rish691 — APPROVED** [review]

**Review by tim-mwangi — APPROVED** [review]
